File: core/src/main/scala/com/graphbrain/eco/Contexts.scala
package com.graphbrain.eco

import scala.collection.mutable
import com.graphbrain.eco.nodes.RuleNode

class Contexts(val rule: RuleNode,
               val prog: Prog,
               val sentence: Words,
               val depth: Integer) {

  val ctxts = mutable.ListBuffer[Context]()

  private val addCtxts = mutable.ListBuffer[Context]()
  private val remCtxts = mutable.ListBuffer[Context]()

  def addContext(c: Context) = addCtxts += c
  def remContext(c: Context) = remCtxts += c

  def applyChanges() = {
    for (c <- addCtxts) ctxts += c
    for (c <- remCtxts) ctxts -= c
    addCtxts.clear()
    remCtxts.clear()
  }

  override def toString = {
    val sb = new mutable.StringBuilder()
    sb.append("Contexts: " + sentence)
    for (c <- ctxts)
      sb.append(c)
    sb.toString()
  }
}

object Contexts {
  def apply(rule: RuleNode, prog: Prog, w: Words, depth: Integer) =
    new Contexts(rule, prog, w, depth)
}

File: src/main/java/com/graphbrain/eco/Contexts.java
package com.graphbrain.eco;

import com.graphbrain.eco.nodes.RuleNode;

import java.util.LinkedList;
import java.util.List;

public class Contexts {
    private Words sentence;
    private RuleNode rule;
    private Prog prog;
    private int depth;
    private List<Context> ctxts;
    private List<Context> addCtxts;
    private List<Context> remCtxts;

    private Context globals;

    public Contexts(RuleNode rule, Prog prog, Words inSentence, int depth) {
        this.rule = rule;
        this.prog = prog;
        this.depth = depth;
        sentence = inSentence.removeFullStop();
        ctxts = new LinkedList<>();
        addCtxts = new LinkedList<>();
        remCtxts = new LinkedList<>();
        globals = null;
    }

    public void addContext(Context c) {addCtxts.add(c);}
    public void remContext(Context c) {remCtxts.add(c);}

    public void applyChanges() {
        for (Context c : addCtxts)
            ctxts.add(c);
        for (Context c : remCtxts)
            ctxts.remove(c);

        addCtxts.clear();
        remCtxts.clear();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(100);
        sb.append("Contexts: ");
        sb.append(sentence);
        for (Context c : ctxts)
            sb.append(c);

        return sb.toString();
    }

    public Words getSentence() {
        return sentence;
    }

    public RuleNode getRule() {
        return rule;
    }

    public Prog getProg() {
        return prog;
    }

    public int getDepth() {
        return depth;
    }

    public List<Context> getCtxts() {
        return ctxts;
    }

    public Context getGlobals() {
        return globals;
    }

    public void setGlobals(Context globals) {
        this.globals = globals;
    }
}